tests/unit-core/test-source-info.c

```diff
@@ -0,0 +1,162 @@
+  TEST_INIT ();
+
+  jerry_init (JERRY_INIT_EMPTY);
+
+  if (!jerry_is_feature_enabled (JERRY_FEATURE_FUNCTION_TO_STRING))
+  {
+    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Source code is not stored!\n");
+    jerry_cleanup ();
+    return 0;
+  }
+
+  jerry_value_t value = jerry_create_null ();
+  jerry_source_info_t *source_info_p = jerry_get_source_info (value);
+  TEST_ASSERT (source_info_p == NULL);
+  jerry_free_source_info (source_info_p);
+  jerry_release_value (value);
+
+  value = jerry_create_object ();
+  source_info_p = jerry_get_source_info (value);
+  TEST_ASSERT (source_info_p == NULL);
+  jerry_free_source_info (source_info_p);
+  jerry_release_value (value);
+
+  jerry_parse_options_t parse_options;
+  const char *source_p = TEST_STRING_LITERAL ("var a = 6");
+
+  value = jerry_parse ((jerry_char_t *) source_p, strlen (source_p), NULL);
+  source_info_p = jerry_get_source_info (value);
+  TEST_ASSERT (source_info_p != NULL);
+  TEST_ASSERT (source_info_p->enabled_fields == JERRY_SOURCE_INFO_HAS_SOURCE_CODE);
+  compare_string (source_info_p->source_code, source_p);
+  TEST_ASSERT (jerry_value_is_undefined (source_info_p->function_arguments));
+  TEST_ASSERT (source_info_p->source_range_start == 0);
+  TEST_ASSERT (source_info_p->source_range_length == 0);
+  jerry_free_source_info (source_info_p);
+  jerry_release_value (value);
+
+  if (jerry_is_feature_enabled (JERRY_FEATURE_MODULE))
+  {
+    parse_options.options = JERRY_PARSE_MODULE;
+
+    value = jerry_parse ((jerry_char_t *) source_p, strlen (source_p), &parse_options);
+
+    jerry_value_t result = jerry_module_link (value, NULL, NULL);
+    TEST_ASSERT (!jerry_value_is_error (result));
+    jerry_release_value (result);
+
+    source_info_p = jerry_get_source_info (value);
+    TEST_ASSERT (source_info_p != NULL);
+    TEST_ASSERT (source_info_p->enabled_fields == JERRY_SOURCE_INFO_HAS_SOURCE_CODE);
+    compare_string (source_info_p->source_code, source_p);
+    TEST_ASSERT (jerry_value_is_undefined (source_info_p->function_arguments));
+    TEST_ASSERT (source_info_p->source_range_start == 0);
+    TEST_ASSERT (source_info_p->source_range_length == 0);
+    jerry_free_source_info (source_info_p);
+
+    result = jerry_module_evaluate (value);
+    TEST_ASSERT (!jerry_value_is_error (result));
+    jerry_release_value (result);
+
+    /* Byte code is released after a successful evaluation. */
+    source_info_p = jerry_get_source_info (value);
+    TEST_ASSERT (source_info_p == NULL);
+    jerry_free_source_info (source_info_p);
+    jerry_release_value (value);
+  }
+
+  source_p = TEST_STRING_LITERAL ("( function f() {} )");
+
+  value = jerry_eval ((const jerry_char_t *) source_p, strlen (source_p), 0);
+  source_info_p = jerry_get_source_info (value);
+  TEST_ASSERT (source_info_p != NULL);
+  TEST_ASSERT (source_info_p->enabled_fields == (JERRY_SOURCE_INFO_HAS_SOURCE_CODE
+                                                 | JERRY_SOURCE_INFO_HAS_SOURCE_RANGE));
```

Comment:
Is it possible to have a case where all three "options"  are true? (has source, has source range, has arguments)

Reply:
Currently no.